### PR TITLE
Protocol invariants testing

### PR DESCRIPTION
**Pull request by grandizzy**

### CODE REVIEW

**Review by ith-harvey — APPROVED** [review]
LGTM, src changes are just to facilitate invariant testing

File: foundry.toml [hunk 1]
```diff
@@ -24,3 +24,9 @@ optimizer_runs = 200
 
 [fuzz]
 runs = 300
+
+[invariant]
+runs           = 10     # The number of calls to make in the invariant tests
+depth          = 100    # The number of times to run the invariant tests
+call_override  = false  # Override calls
+fail_on_revert = false  # Fail the test if the contract reverts
```
Comment (EdNoepel) [inline foundry.toml:32]: Contrary to Lucas' recommendation, I agree with this.  Our tests should ensure pool reverts where appropriate.

File: src/base/Pool.sol [hunk 2]
```diff
@@ -715,4 +715,19 @@ abstract contract Pool is Clone, ReentrancyGuard, Multicall, IPool {
...
         );
     }
+
+    /// @inheritdoc IPoolState
+    function totalAuctionsInPool() external view override returns (uint256) {
+        return auctions.noOfAuctions;
+    }
+
+    /// @inheritdoc IPoolState
+    function totalDebt() external view override returns (uint256) {
```
Comment (EdNoepel) [inline src/base/Pool.sol]: Disagree with the naming of this and line 730.  Naming, not just comments, should reveal that debt is in t0 terms.

File: tests/forge/ERC20Pool/invariants/BasicInvariants.t.sol [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+            // tp of a loan at index 'i' in loan array should be greater than equals to loans at index '2i' and '2i+1'
+            (, uint256 tp1) = _pool.loanInfo(2 * loanId);
+            (, uint256 tp2) = _pool.loanInfo(2 * loanId + 1);
+
+            require(tp >= tp1, "Incorrect loan heap");
+            require(tp >= tp2, "Incorrect loan heap");
+        }
+    }
+
+    // int-erest should only update once in 12 hours
```
Comment (EdNoepel) [inline tests/forge/ERC20Pool/invariants/BasicInvariants.t.sol]: nit: errant hyphen in comment

File: tests/forge/ERC20Pool/invariants/handlers/BasicPoolHandler.sol [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+        try _pool.removeQuoteToken(amount, bucketIndex) {
+            // Post condition
+            (uint256 lpBalanceAfter, ) = _pool.lenderInfo(bucketIndex, _actor);
+            require(lpBalanceAfter < lpBalanceBefore, "LP balance should decrease");
+            shouldExchangeRateChange = false;
+            shouldReserveChange      = false;
+        }
+        catch (bytes memory _err){
+            bytes32 err = keccak256(_err);
+            require(err == keccak256(abi.encodeWithSignature("LUPBelowHTP()")) || err == keccak256(abi.encodeWithSignature("InsufficientLiquidity()")) || err == keccak256(abi.encodeWithSignature("RemoveDepositLockedByAuctionDebt()")));
```
Comment (EdNoepel) [inline tests/forge/ERC20Pool/invariants/handlers/BasicPoolHandler.sol]: Could also raise `NoClaim` here.  Is it only trying to capture exceptions expected by the test?
nit: would be more readable if each error was on its own line

File: tests/forge/ERC20Pool/invariants/handlers/BasicPoolHandler.sol [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+
+        _pool.removeCollateral(amount, bucketIndex);
+
+        // Post condition
+        (uint256 lpBalanceAfter, ) = _pool.lenderInfo(bucketIndex, _actor);
+        require(lpBalanceAfter < lpBalanceBefore, "LP balance should decrease");
+    }
+
+    /**************************************************************************************************************************************/
+    /*** Borrower Functions                                                                                                               ***/
```
Comment (EdNoepel) [inline tests/forge/ERC20Pool/invariants/handlers/BasicPoolHandler.sol]: nit: comment alignment

File: tests/forge/ERC20Pool/invariants/handlers/BasicPoolHandler.sol [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+            addQuoteToken(amount, fromIndex);
+        }
+
+        try _pool.moveQuoteToken(amount, fromIndex, toIndex, block.timestamp + 1 minutes) {
+            shouldExchangeRateChange = false;
+            shouldReserveChange      = false;
+        }
+        catch (bytes memory _err){
+            bytes32 err = keccak256(_err);
+            require(err == keccak256(abi.encodeWithSignature("LUPBelowHTP()")) || err == keccak256(abi.encodeWithSignature("InsufficientLiquidity()")) || err == keccak256(abi.encodeWithSignature("RemoveDepositLockedByAuctionDebt()")));
```
Comment (EdNoepel) [inline tests/forge/ERC20Pool/invariants/handlers/BasicPoolHandler.sol]: `LUPBelowHTP` and `InsufficientLiquidity` make sense here, but I don't see where `RemoveDepositLockedByAuctionDebt` could be raised.  Expect `MoveToSamePrice`, `DustAmountNotExceeded`, `InvalidIndex`, `BucketBankruptcyBlock` too.

File: tests/forge/ERC20Pool/invariants/handlers/BasicPoolHandler.sol (hunk 5 above)
Comment (EdNoepel) [reply tests/forge/ERC20Pool/invariants/handlers/BasicPoolHandler.sol]: same on line 298

File: tests/forge/ERC20Pool/invariants/handlers/IBaseHandler.sol [hunk 7]
```diff
@@ -0,0 +1,23 @@
+// SPDX-License-Identifier: UNLICENSED
+
+pragma solidity 0.8.14;
+
+// TODO: Complete this interface
+interface IBaseHandler {
+
+    function getActorsCount() external view returns(uint256);
+
+    function _actors(uint256) external view returns(address);
```
Comment (EdNoepel) [inline tests/forge/ERC20Pool/invariants/handlers/IBaseHandler.sol]: If they are all external views, why should this one have an underscore prefix?

File: tests/forge/ERC20Pool/invariants/handlers/IBaseHandler.sol [hunk 8]
```diff
@@ -0,0 +1,23 @@
+// SPDX-License-Identifier: UNLICENSED
+
+pragma solidity 0.8.14;
+
+// TODO: Complete this interface
```
Comment (EdNoepel) [inline tests/forge/ERC20Pool/invariants/handlers/IBaseHandler.sol]: Would be better to briefly state what is missing from interface.

File: src/interfaces/pool/commons/IPoolState.sol [hunk 9]
```diff
@@ -222,6 +222,24 @@ interface IPoolState {
...
 
+    /**
+     *  @notice Returns the total number of active auctions in pool
+     *  @return totalAuctions_ number of active auctions.
+     */
+    function totalAuctionsInPool() external view returns (uint256);
+
+     /**
+     *  @notice Returns the `t0Debt` state variable.
+     *  @return The total t0Debt in the system, in WAD units.
```
Comment (EdNoepel) [inline src/interfaces/pool/commons/IPoolState.sol:234]: Here and 239, comments should mention this needs to be multiplied by inflator to calculate current debt.

**Review by EdNoepel — COMMENTED** [review]
Good stuff.  Several comments inline, but my only real concern is the naming of new high-visibility `Pool` members which expose debt.

File: src/interfaces/pool/commons/IPoolState.sol (hunk 9 above)
Comment (grandizzy) [reply src/interfaces/pool/commons/IPoolState.sol:234]: done in https://github.com/ajna-finance/contracts/pull/609/commits/73be658e9ff3f25655784990777916fd1ebc504f

File: src/base/Pool.sol (hunk 2 above)
Comment (grandizzy) [reply src/base/Pool.sol]: changed in https://github.com/ajna-finance/contracts/pull/609/commits/73be658e9ff3f25655784990777916fd1ebc504f

File: tests/forge/ERC20Pool/invariants/BasicInvariants.t.sol (hunk 3 above)
Comment (grandizzy) [reply tests/forge/ERC20Pool/invariants/BasicInvariants.t.sol]: fixed in https://github.com/ajna-finance/contracts/pull/609/commits/73be658e9ff3f25655784990777916fd1ebc504f

File: tests/forge/ERC20Pool/invariants/handlers/BasicPoolHandler.sol (hunk 4 above)
Comment (grandizzy) [reply tests/forge/ERC20Pool/invariants/handlers/BasicPoolHandler.sol]: added in https://github.com/ajna-finance/contracts/pull/609/commits/73be658e9ff3f25655784990777916fd1ebc504f

File: tests/forge/ERC20Pool/invariants/handlers/BasicPoolHandler.sol (hunk 5 above)
Comment (grandizzy) [reply tests/forge/ERC20Pool/invariants/handlers/BasicPoolHandler.sol]: changed in https://github.com/ajna-finance/contracts/pull/609/commits/73be658e9ff3f25655784990777916fd1ebc504f

File: tests/forge/ERC20Pool/invariants/handlers/BasicPoolHandler.sol (hunk 6 above)
Comment (grandizzy) [reply tests/forge/ERC20Pool/invariants/handlers/BasicPoolHandler.sol]: remoived in https://github.com/ajna-finance/contracts/pull/609/commits/73be658e9ff3f25655784990777916fd1ebc504f

File: tests/forge/ERC20Pool/invariants/handlers/IBaseHandler.sol (hunk 7 above)
Comment (grandizzy) [reply tests/forge/ERC20Pool/invariants/handlers/IBaseHandler.sol]: changed in https://github.com/ajna-finance/contracts/pull/609/commits/73be658e9ff3f25655784990777916fd1ebc504f

File: tests/forge/ERC20Pool/invariants/handlers/IBaseHandler.sol (hunk 8 above)
Comment (grandizzy) [reply tests/forge/ERC20Pool/invariants/handlers/IBaseHandler.sol]: don't know exactly what else is needed, I just removed TODO for now

**Review by EdNoepel — APPROVED** [review]
LGTM!

**Review by mattcushman — APPROVED** [review]
lgtm -- great work!

**Pull request merged**